x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/execution_plan/required_permissions.ts

```diff
@@ -0,0 +1,153 @@
+
+export interface RequiredPermissions {
+  cluster: string[];
+  index: Record<string, string[]>;
+}
+
+export function mergeRequiredPermissions(permissions: RequiredPermissions[]): RequiredPermissions {
+  const clusterSet = new Set<string>();
+  const indexPermissions: Record<string, Set<string>> = {};
+
+  permissions.forEach((permission) => {
+    // Add cluster permissions to the set
+    permission.cluster.forEach((perm) => clusterSet.add(perm));
+
+    // Process index permissions
+    Object.entries(permission.index).forEach(([index, perms]) => {
+      if (!indexPermissions[index]) {
+        indexPermissions[index] = new Set<string>();
+      }
+      perms.forEach((perm) => indexPermissions[index].add(perm));
+    });
+  });
+
+  // Convert sets back to arrays for the return value
+  const result: RequiredPermissions = {
+    cluster: Array.from(clusterSet),
+    index: {},
+  };
+
+  Object.entries(indexPermissions).forEach(([index, permSet]) => {
+    result.index[index] = Array.from(permSet);
+  });
+
+  return result;
+}
+
+export function getRequiredPermissionsForActions(actions: ActionsByType): RequiredPermissions {
+  const permissions: RequiredPermissions[] = [];
+
+  if (actions.upsert_component_template.length > 0) {
+    permissions.push({
+      cluster: ['manage_index_templates'],
+      index: {},
+    });
+  }
+
+  if (actions.delete_component_template.length > 0) {
+    permissions.push({
+      cluster: ['manage_index_templates'],
+      index: {},
+    });
+  }
+
+  if (actions.upsert_index_template.length > 0) {
+    permissions.push({
+      cluster: ['manage_index_templates'],
+      index: {},
+    });
+  }
+
+  if (actions.delete_index_template.length > 0) {
+    permissions.push({
+      cluster: ['manage_index_templates'],
+      index: {},
+    });
+  }
+
+  if (actions.upsert_ingest_pipeline.length > 0) {
+    permissions.push({
+      cluster: ['manage_pipeline'],
+      index: {},
+    });
+  }
+
+  if (actions.delete_ingest_pipeline.length > 0) {
+    permissions.push({
+      cluster: ['manage_pipeline'],
+      index: {},
+    });
+  }
+
+  if (actions.append_processor_to_ingest_pipeline.length > 0) {
+    permissions.push({
+      cluster: ['manage_pipeline'],
+      index: {},
+    });
+  }
+
+  if (actions.delete_processor_from_ingest_pipeline.length > 0) {
+    permissions.push({
+      cluster: ['manage_pipeline'],
+      index: {},
+    });
+  }
+
+  if (actions.upsert_datastream.length > 0) {
+    const indexPermissions: Record<string, string[]> = {};
+    actions.upsert_datastream.forEach((action) => {
+      indexPermissions[action.request.name] = ['create_index'];
+    });
+    permissions.push({
+      cluster: [],
+      index: indexPermissions,
+    });
+  }
+
+  if (actions.update_lifecycle.length > 0) {
+    const indexPermissions: Record<string, string[]> = {};
+    actions.update_lifecycle.forEach((action) => {
+      indexPermissions[action.request.name] = [
+        'manage_ilm',
```

Comment:
[manage_ilm](https://www.elastic.co/docs/deploy-manage/users-roles/cluster-or-deployment-auth/elasticsearch-privileges) is not available in serverless so we should likely include it conditionally

Reply:
Not necessary

```
POST _security/user/_has_privileges
{
    "cluster": ["manage_ilm"]
}

->
#! HasPrivileges request includes privileges for features not available in serverless mode; you will not have access to these features regardless of your permissions; cluster privileges: [manage_ilm];
{
  "username": "----",
  "has_all_requested": true,
  "cluster": {
    "manage_ilm": true
  },
  "index": {},
  "application": {}
}
```